lib/spack/spack/spec.py

```diff
@@ -2232,14 +2232,25 @@ def concretize(self, tests=False):
             mvar._patches_in_order_of_appearance = list(
                 t[-1] for t in ordered_hashes)
 
+        # This is awful!!!!
+        # TODO: anything better
+        # This allows us to access the package class for externals that
+        # override their prefix. We then mark the spec as concrete again after
+        # setting the external paths.
+        self._mark_concrete()
```

Comment:
why is this needed?  you don't need the specs to be concrete just to get the package - is there something I'm missing here?

Reply:
`Spec.package` caches the value of package, and instantiates it with the spec. I think in this case we might be able to get away with it since we know we're about to mark it concrete, but I have to think through the details of that.

When I started the branch we had an assert at the top of `Spec.package`, but that's apparently outdated.

Reply:
Looks like this is gone.  marking resolved.